File: MovieDB/app/src/main/java/com/tp/projects/moviedb/movies/MovieFragment.java
package com.tp.projects.moviedb.movies;


import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.tp.projects.moviedb.R;
import com.tp.projects.moviedb.util.GeneralRetrofitResponseHandler;
import com.tp.projects.moviedb.util.JSONParser;
import com.tp.projects.moviedb.util.MovieDBComponentInjector;
import com.tp.projects.moviedb.util.NetworkHandler;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * A simple {@link Fragment} subclass.
 */
public class MovieFragment extends Fragment {

  @Inject
  NetworkHandler networkHandler;
  private Context ctx;
  private static List<MovieData> movieList;
  private List<MovieData> searchedList;
  private View mainView;

  public MovieFragment() {
    if (getArguments() != null) {
    String query = getArguments().getString("searchedList");
      if(query != null) {
        searchedList = new ArrayList<>();
        for (MovieData movie : movieList) {
          if (movie.getTitle().contains(query)) {
            searchedList.add(movie);
          }
        }
      }
    }
  }


  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    MovieDBComponentInjector.instance().inject(this);
    ctx = getActivity();
    networkHandler.downloadMovieDataRetrofit()
      .subscribeOn(Schedulers.io())
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe(new GeneralRetrofitResponseHandler(getActivity()) {
        @Override
        public void onNext(JsonElement jsonElement) {
          parseMovieJSONData(jsonElement.getAsJsonObject());
          initializeTileLayout();
        }

        @Override
        public void onCompleted() {
        }
      });

  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
                           Bundle savedInstanceState) {
    mainView = inflater.inflate(R.layout.fragment, container, false);
    return mainView;
  }

  private void parseMovieJSONData(JsonObject result) {
    movieList = new ArrayList<>();
    JsonArray jsonList = result.getAsJsonArray("results");
    for (JsonElement showJSON : jsonList) {
      MovieData movie = (MovieData) JSONParser.returnParsedClass(showJSON, MovieData.class);
      setImageURLs(movie);
      movieList.add(movie);
    }
  }

  private void initializeTileLayout() {
    RecyclerView recyclerView = (RecyclerView) mainView.findViewById(R.id.tiles_container);
    if (recyclerView != null) {
      recyclerView.setHasFixedSize(true);
      recyclerView.setLayoutManager(new LinearLayoutManager(ctx));
      if (searchedList != null) {
        recyclerView.setAdapter(new MovieTilesAdapter(ctx, searchedList));
      } else {
        recyclerView.setAdapter(new MovieTilesAdapter(ctx, movieList));
      }
    }
  }

  public void setImageURLs(MovieData movie) {
    movie.setPosterPath(networkHandler.createTileImageURL(movie.getPosterPath()));
    movie.setBackdropPath(networkHandler.createHeaderImageURL(movie.getBackdropPath()));
  }
}

File: MovieDB/app/src/main/java/com/tp/projects/moviedb/movies/MovieFragment.kt
package com.tp.projects.moviedb.movies

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.tp.projects.moviedb.R
import com.tp.projects.moviedb.util.NetworkHandler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import org.koin.android.ext.android.inject
import timber.log.Timber

class MovieFragment : Fragment() {

    private val networkHandler: NetworkHandler by inject()

    private val disposeBag = CompositeDisposable()

    private lateinit var mainView: View

    private var searchedList = listOf<MovieData>()
    private var movieList = mutableListOf<MovieData>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        arguments?.getString("searchedList")?.let { word ->
            searchedList = movieList.filter { it.title.contains(word) }
        }

        networkHandler.downloadMovieDataRetrofit()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ movies ->
                    //Todo: Need to refactor URL creation
                    movieList.addAll(movies.results.map { it.apply { setImageURLs(it) } })
                    arguments?.getString("searchedList")?.let { word ->
                        searchedList = movieList.filter {
                            val lowCaseTitle = it.title.toLowerCase()
                            lowCaseTitle.contains(word)
                        }
                    }
                    initializeTileLayout(requireContext())
                }, { Timber.e(it) })
                .apply { disposeBag.add(this) }
    }

    override fun onDestroy() {
        super.onDestroy()
        disposeBag.clear()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        mainView = inflater.inflate(R.layout.fragment, container, false)
        return mainView
    }

    private fun initializeTileLayout(ctx: Context) {
        mainView.findViewById<RecyclerView>(R.id.tiles_container).apply {
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(ctx)
            adapter = if (searchedList.isEmpty()) {
                MovieTilesAdapter(ctx, movieList)
            } else {
                MovieTilesAdapter(ctx, searchedList)
            }
        }
    }

    private fun setImageURLs(movie: MovieData) {
        movie.posterPath = networkHandler.createTileImageURL(movie.posterPath)
        movie.backdropPath = networkHandler.createHeaderImageURL(movie.backdropPath)
    }

}
